### PR TITLE
fsmonitor: don't fill bitmap with removed entries

### PR DESCRIPTION
**Pull request by wilbaker**
`fill_fsmonitor_bitmap` was populating `fsmonitor_dirty` with all entries in the cache, even those that have been marked for removal.  This caused a problem when subsequent git commands read the index, as the bit's position in `fsmonitor_dirty` would no longer match the index in the cache.  Additionally, the number of entries in `fsmonitor_dirty` could exceed the number of entries in the cache.  If the number of entries in `fsmonitor_dirty` exceeded the number of entries in the cache it resulted in a segfault when `tweak_fsmonitor` was called because
`ewah_each_bit` would go beyond the bounds of the cache.

### CODE REVIEW

**Comment by wilbaker** [comment]
All checks passed with a base of `vfs-2.23.0`, just realized I want to target `features/sparse-checkout-2.23.0`

File: fsmonitor.c [hunk 1]
```diff
@@ -56,11 +57,13 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
...
-	unsigned int i;
+	unsigned int i, removed = 0;
 	istate->fsmonitor_dirty = ewah_new();
 	for (i = 0; i < istate->cache_nr; i++)
-		if (!(istate->cache[i]->ce_flags & CE_FSMONITOR_VALID))
-			ewah_set(istate->fsmonitor_dirty, i);
+		if (istate->cache[i]->ce_flags & CE_REMOVE)
+			removed++;
+		else if (!(istate->cache[i]->ce_flags & CE_FSMONITOR_VALID))
+			ewah_set(istate->fsmonitor_dirty, i - removed);
```
Comment (wilbaker) [inline fsmonitor.c]: Assuming the change looks good and passes existing tests, are there new tests we can add?

File: fsmonitor.c [hunk 2]
```diff
@@ -15,7 +15,8 @@ static void fsmonitor_ewah_callback(size_t pos, void *is)
 {
 	struct index_state *istate = (struct index_state *)is;
 	struct cache_entry *ce = istate->cache[pos];
-
+
```
Comment (jeffhostetler) [inline fsmonitor.c]: I can’t see the whole context on my iPad, but I’d consider changing this a bit. 

assert() is a debug-only macro, so it’s wont give production protection.  Suggest “if (pos < …) BUG(…);” instead.  Or maybe just “return” if that’s more appropriate.

Also, maybe split the declaration and assignment of “ce”,  but then again as long as you don’t deref the pointer, that’s more of a nit.

File: fsmonitor.c (hunk 1 above)
Comment (dscho) [reply fsmonitor.c]: Absolutely! The best place would be [t/t7519-status-fsmonitor.sh](https://github.com/microsoft/git/blob/vfs-2.23.0/t/t7519-status-fsmonitor.sh), I think.

So I played around with this a bit, and it seems as if the `git restore` call (or the `git checkout` call, if you're old school) is crucial for this bug to reproduce. This patch seems to work for me: reverting your change to `fsmonitor.c` causes the test case to fail due to a segmentation fault, reinstating the changes fixes it again:

```diff
diff --git a/fsmonitor.c b/fsmonitor.c
index 231e83a94db..0119eda37ce 100644
--- a/fsmonitor.c
+++ b/fsmonitor.c
@@ -56,11 +56,15 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 
 void fill_fsmonitor_bitmap(struct index_state *istate)
 {
-	unsigned int i;
+	unsigned int i, removed = 0;
 	istate->fsmonitor_dirty = ewah_new();
-	for (i = 0; i < istate->cache_nr; i++)
-		if (!(istate->cache[i]->ce_flags & CE_FSMONITOR_VALID))
-			ewah_set(istate->fsmonitor_dirty, i);
+	for (i = 0; i < istate->cache_nr; i++) {
+		unsigned int flags = istate->cache[i]->ce_flags;
+		if (flags & CE_REMOVE)
+			removed++;
+		else if (!(flags & CE_FSMONITOR_VALID))
+			ewah_set(istate->fsmonitor_dirty, i - removed);
+	}
 }
 
 void write_fsmonitor_extension(struct strbuf *sb, struct index_state *istate)
diff --git a/t/t7519-status-fsmonitor.sh b/t/t7519-status-fsmonitor.sh
index 81a375fa0ff..9e9621a3548 100755
--- a/t/t7519-status-fsmonitor.sh
+++ b/t/t7519-status-fsmonitor.sh
@@ -354,4 +354,15 @@ test_expect_success 'discard_index() also discards fsmonitor info' '
 	test_cmp expect actual
 '
 
+test_expect_success 'fsmonitor extension ignores removed index entries' '
+	test_commit initial &&
+	removed=$(seq 1 100 | sed "s/^/z/") &&
+	touch $removed &&
+	git add $removed &&
+	test_config core.fsmonitor "$TEST_DIRECTORY/t7519/fsmonitor-env" &&
+	FSMONITOR_LIST="$removed" git restore -S $removed &&
+	FSMONITOR_LIST="$removed" git status
+'
+
 test_done
diff --git a/t/t7519/fsmonitor-env b/t/t7519/fsmonitor-env
new file mode 100755
index 00000000000..8f1f7ab164e
--- /dev/null
+++ b/t/t7519/fsmonitor-env
@@ -0,0 +1,24 @@
+#!/bin/sh
+#
+# An test hook script to integrate with git to test fsmonitor.
+#
+# The hook is passed a version (currently 1) and a time in nanoseconds
+# formatted as a string and outputs to stdout all files that have been
+# modified since the given time. Paths must be relative to the root of
+# the working tree and separated by a single NUL.
+#
+#echo "$0 $*" >&2
+
+if test "$#" -ne 2
+then
+	echo "$0: exactly 2 arguments expected" >&2
+	exit 2
+fi
+
+if test "$1" != 1
+then
+	echo "Unsupported core.fsmonitor hook version." >&2
+	exit 1
+fi
+
+printf '%s\n' $FSMONITOR_LIST
```

Comment (jeffhostetler) [reply fsmonitor.c]: Sorry if this is a dumb question (ipad) but are there any places where it would be helpful to assert the length of the bitmap and the cache entry array are equal?  Just trying to verify that any code that manipulates one also updates the other.

File: fsmonitor.c [hunk 3]
```diff
@@ -56,11 +57,13 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
...
 void fill_fsmonitor_bitmap(struct index_state *istate)
 {
-	unsigned int i;
+	unsigned int i, removed = 0;
 	istate->fsmonitor_dirty = ewah_new();
 	for (i = 0; i < istate->cache_nr; i++)
-		if (!(istate->cache[i]->ce_flags & CE_FSMONITOR_VALID))
-			ewah_set(istate->fsmonitor_dirty, i);
+		if (istate->cache[i]->ce_flags & CE_REMOVE)
+			removed++;
```
[the remark this answers is not in the document]
Comment (wilbaker) [reply fsmonitor.c]: @dscho I am good with either implementation 😄   

Do you have a feel for what would be more upstreamable?  For what it's worth I went with the current implementation in an attempt to match the check in `do_write_index`:

```
	for (i = removed = extended = 0; i < entries; i++) {
		if (cache[i]->ce_flags & CE_REMOVE)
			removed++;
```
I was also thinking I'd add a small comment giving some explanation of why this check is there (let me know if you think that'd be a bad idea).

File: fsmonitor.c (hunk 1 above)
Comment (wilbaker) [reply fsmonitor.c]: > are there any places where it would be helpful to assert the length of the bitmap and the cache entry array are equal? Just trying to verify that any code that manipulates one also updates the other.

@dscho please correct me if my understanding is wrong.  

Based on my reading of the code the bitmap length should always be <= to the length of the cache entry array, but it won't necessarily be equal to the length of the cache entries.  `ewah_set` grows the bitmap only when needed, and so if the last X cache entries have the `CE_FSMONITOR_VALID` flag set, the length of `fsmonitor_dirty` will be `istate->cache_nr - X`.

I will see if there is anywhere else that `fsmonitor_dirty` is being adjusted where an `assert` that its length is less than `istate->cache_nr` is appropriate.

File: fsmonitor.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (wilbaker) [reply fsmonitor.c]: > Do you have a feel for what would be more upstreamable? 

After taking a closer look, I prefer that your option keeps the current index into `fsmonitor_dirty` @dscho, I will update this.

File: fsmonitor.c (hunk 1 above)
Comment (wilbaker) [reply fsmonitor.c]: > I will see if there is anywhere else that fsmonitor_dirty is being adjusted where an assert that its length is less than istate->cache_nr is appropriate.

I've found a few spots (maybe too many?) and will include them in my next push, please let me know what you think.

**Comment by wilbaker** [comment]
@dscho @jeffhostetler thanks for all the great feedback so far.

I am still working on updating the tests (and will include them in a new commit), but I've pushed a cleanup of the product code based on feedback so far.  I'm a bit uncertain if I added too many new asserts, and so please let me know your thoughts on that in particular.

File: fsmonitor.c (hunk 1 above)
Comment (wilbaker) [reply fsmonitor.c]: @dscho thanks for writing up a test!  For my own understanding, is this:

`sed "s/^/z/"`

To ensure the entries being added/removed appear at the end of the cache?

File: t/t7519-status-fsmonitor.sh [hunk 4]
```diff
@@ -354,4 +354,18 @@ test_expect_success 'discard_index() also discards fsmonitor info' '
...
+test_expect_success 'compare status before and after staging/unstaging ' '
+	test_commit initial &&
+	removed=$(seq 1 100 | sed "s/^/z/") &&
+	touch $removed &&
+	git status >expected &&
+	git add $removed &&
+	test_config core.fsmonitor "$TEST_DIRECTORY/t7519/fsmonitor-env" &&
+	FSMONITOR_LIST="$removed" git restore -S $removed &&
+	FSMONITOR_LIST="$removed" git status >actual &&
+	test_cmp expected actual
```
Comment (wilbaker) [inline t/t7519-status-fsmonitor.sh]: @dscho I made a few tweaks here, please let me know what you think.  

I changed the description and had the test validate `status` because I wasn't sure the best way to demonstrate that the test was actually validating that fsmonitor was ignoring the removed entries.

I confirmed the test fails without the fix.

File: fsmonitor.c [hunk 5]
```diff
@@ -50,17 +55,29 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 	}
 	istate->fsmonitor_dirty = fsmonitor_dirty;
 
+	assert(istate->fsmonitor_dirty->bit_size <= istate->cache_nr);
```
Comment (dscho) [inline fsmonitor.c]: Maybe make that a `BUG()`, too? I can easily imagine this to help stave off hard-to-debug problems in the future.

File: fsmonitor.c [hunk 6]
```diff
@@ -50,17 +55,29 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
...
+		switch (flags & (CE_REMOVE | CE_FSMONITOR_VALID)) {
+		case 0:
+			ewah_set(istate->fsmonitor_dirty, j);
+			/* fallthru */
+		case CE_FSMONITOR_VALID:
+			j++;
+		default:
+			/* do not include entries being removed in fsmonitor_dirty */
+			break;
+		}
```
Comment (dscho) [inline fsmonitor.c]: Heh. Now that I read this, I actually prefer your version, I now find it way more readable...

If you want to keep the `j` way, maybe this?

```c
if (flags & CE_REMOVE)
    continue;
if (!(flags & CE_FSMONITOR_VALID))
    ewah_set(istate->fsmonitor_dirty, j);
j++;
```

Comment (dscho) [reply fsmonitor.c]: But now that I think of it even more, I don't think the `j` way would be more upstreamable. I'd think that the way you had it before would make more sense, maybe using `skipped` instead of `removed`?

File: fsmonitor.c [hunk 7]
```diff
@@ -71,6 +88,8 @@ void write_fsmonitor_extension(struct strbuf *sb, struct index_state *istate)
 	uint32_t ewah_size = 0;
 	int fixup = 0;
 
+	assert(istate->fsmonitor_dirty->bit_size <= istate->cache_nr);
```
Comment (dscho) [inline fsmonitor.c]: Again, a `BUG()` might make it easier to catch these things early, should they hit production.

File: fsmonitor.c [hunk 8]
```diff
@@ -236,6 +255,7 @@ void tweak_fsmonitor(struct index_state *istate)
 			}
 
 			/* Mark all previously saved entries as dirty */
+			assert(istate->fsmonitor_dirty->bit_size <= istate->cache_nr);
```
Comment (dscho) [inline fsmonitor.c]: And again, I would prefer this as a `BUG()`.

File: t/t7519-status-fsmonitor.sh (hunk 4 above)
Comment (dscho) [reply t/t7519-status-fsmonitor.sh]: Yes, you introduced the `expected` vs `actual` test. I like it! To keep with the convention, could you rename it to `expect`? That's how it is called in most places in Git's test suite.

File: fsmonitor.c (hunk 1 above)
Comment (dscho) [reply fsmonitor.c]: Yes.

File: fsmonitor.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (dscho) [reply fsmonitor.c]: Your version has the advantage of not incrementing two variables _all the time_ ;-)

File: fsmonitor.c (hunk 6 above)
Comment (wilbaker) [reply fsmonitor.c]: >  I'd think that the way you had it before would make more sense, maybe using `skipped` instead of `removed`?

I will switch this back and I think `skipped` is a better name than `removed`, thanks!

File: t/t7519-status-fsmonitor.sh (hunk 4 above)
Comment (wilbaker) [reply t/t7519-status-fsmonitor.sh]: @dscho I hit a failure on `osx-clang` after pushing my latest set of changes (the other platforms were fine), here is the diff of `git-status`:

```
2019-09-30T17:47:34.6313770Z ++ diff -u expect actual
2019-09-30T17:47:34.6314380Z --- expect	2019-09-30 17:42:28.000000000 +0000
2019-09-30T17:47:34.6315030Z +++ actual	2019-09-30 17:42:28.000000000 +0000
2019-09-30T17:47:34.6316420Z @@ -11,10 +11,6 @@
2019-09-30T17:47:34.6316690Z    (use "git add <file>..." to include in what will be committed)
2019-09-30T17:47:34.6316900Z  	after
2019-09-30T17:47:34.6317070Z  	before
2019-09-30T17:47:34.6317580Z -	dir1/new
2019-09-30T17:47:34.6318160Z -	dir1/untracked
2019-09-30T17:47:34.6318720Z -	dir2/new
2019-09-30T17:47:34.6319280Z -	dir2/untracked
2019-09-30T17:47:34.6319820Z  	dot-git/
2019-09-30T17:47:34.6320020Z  	new
2019-09-30T17:47:34.6320540Z  	trace-after
2019-09-30T17:47:34.6320760Z error: last command exited with $?=1
```
https://gvfs.visualstudio.com/23a36a33-54f4-42be-84ae-6286f1f5fd6a/_apis/build/builds/16835/logs/160

Any guesses as to what might have caused it? (the `dir1` and `dir2` files are created by a different test).

This is a bit strange because:
- The tests passed fine last week: https://gvfs.visualstudio.com/ci/_build/results?buildId=16778&view=results
- All of the test files that this test uses appear well after the `dir` files in the index
- The test passed when I re-ran it: https://gvfs.visualstudio.com/ci/_build/results?buildId=16835&view=logs&j=bb92627a-0f17-5507-454a-87285e0fd6b8 and I'm having trouble seeing what would make this test racy.  

I did see that the time timing of when t7519 ran was different across runs, and I'm not sure if that could be related.

Do you have any advice on how to proceed?  I have been unable to repro this failure locally and thus far have only seen it once on the build machines.

cc: @derrickstolee @jeffhostetler for their thoughts as well

Comment (jeffhostetler) [reply t/t7519-status-fsmonitor.sh]: I'm having a little trouble following what the test is doing, so I'm going to be vague here. :-)

On line 363, we have "git status >expect" and on line 367 "FSMONITOR... git status >actual".
This threw me a minute -- I'm was thinking that "expect" would be a test constant. 

Perhaps it would be better to explicitly (hard-code) the correct answer using a here-doc (the
"cat... EOF" thing).  And then have these 2 status commands be "actual1" and "actual2" and
compare each of them against the well-known "expect".   Then at least we'd know which side
was wrong.

There may be a race in here, but I don't know enough about the fsmonitor code to comment
right now.

Comment (wilbaker) [reply t/t7519-status-fsmonitor.sh]: > This threw me a minute -- I'm was thinking that "expect" would be a test constant.
>
>Perhaps it would be better to explicitly (hard-code) the correct answer using a here-doc (the "cat... EOF" thing). And then have these 2 status commands be "actual1" and "actual2" and compare each of them against the well-known "expect". Then at least we'd know which side was wrong.

I was trying to use test code as an example when I made the change:

```
# Status is well tested elsewhere so we'll just ensure that the results are
# the same when using core.fsmonitor.
test_expect_success 'compare status with and without fsmonitor' '
	write_integration_script &&
	clean_repo &&
	dirty_repo &&
	git add new &&
	git add dir1/new &&
	git add dir2/new &&
	git status >actual &&
	git -c core.fsmonitor= status >expect &&
	test_i18ncmp expect actual
```

Although I now see that if we do end up sticking with the current approach that `test_i18ncmp` should be used instead of `test_cmp`.  

The messy part I've found with trying to declare an "expected" constant is the expected output will change base on which of the tests cases have run prior to running this test case (and would it be okay for the expected test case to always assume all previous tests have run)?

I'm wondering if I should revert it to the original proposed by @dscho that did not include a comparison of the output of status.  After trying a bunch of iterations I can confirm that to repro the segfault without the fix in this PR the `test_config core.fsmonitor` line must be set after running the initial `git status` and `git add`, but I have not been able to figure out why we must wait until after those git commands have run.  

In other words, I'm not sure why I can't hit a repro when when I set the `core.fsmonitor` ahead of the initial `git status` and `git add`.  @dscho how did you track down this ordering down?

Related to the earlier discussion regarding expected output, in general I'm having some trouble figuring out how each of these test cases impacts the next.  For example:

- `./t7519-status-fsmonitor.sh --run=29` fails without the fix
- `./t7519-status-fsmonitor.sh --run=1,29` passes without the fix
- `./t7519-status-fsmonitor.sh --run=1,28,29` fails without the fix

Is there any general guidance as to whether all test cases should be able to run individually?  Or should developers assume that all test cases will be run together?

Comment (wilbaker) [reply t/t7519-status-fsmonitor.sh]: > I'm wondering if I should revert it to the original proposed by @dscho that did not include a comparison of the output of status. 

I pushed a second commit with that approach here:

https://github.com/microsoft/git/pull/201/commits/11ff0f5ef30bb635551581af4c27964a4cec5a2e

Comment (jeffhostetler) [reply t/t7519-status-fsmonitor.sh]: I'm not sure about the larger questions you're asking here, but you might look at "test_when_finished" clause that can cleanup per-test cruft and eliminate some of
these dependencies.

File: t/t7519-status-fsmonitor.sh [hunk 9]
```diff
@@ -354,4 +354,16 @@ test_expect_success 'discard_index() also discards fsmonitor info' '
 	test_cmp expect actual
 '
 
+# Use test files that start with 'z' so that the entries being added
+# and removed appear at the end of the index.
+test_expect_success 'status succeeds and after staging/unstaging ' '
```
Comment (derrickstolee) [inline t/t7519-status-fsmonitor.sh]: nit: do you mean to have "and" here? Is it extra, or are you missing a word after?

Comment (derrickstolee) [reply t/t7519-status-fsmonitor.sh]: (Looking at the "Alternate testing approach" commit, it looks like it should be dropped along with the word "before")

Comment (wilbaker) [reply t/t7519-status-fsmonitor.sh]: Nice catch thanks!

Comment (wilbaker) [reply t/t7519-status-fsmonitor.sh]: @derrickstolee I will push a fix for this and squash all commits before merging.  I'm holding the PR right now to hear @dscho's on the other thread regarding this test.

Comment (dscho) [reply t/t7519-status-fsmonitor.sh]: Oh, from my side it looks good, too!
